File: ios/Classes/TexRenderer.swift
//
//  TexGenerator.swift
//  flutter_tex_js
//
//  Created by Aaron Madlon-Kay on 2020/07/01.
//

import Foundation
import WebKit

@available(iOS 11.0, *)
fileprivate let assetsUrl = Bundle(for: TexRenderer.self).url(forResource: "flutter_tex_js_katex", withExtension: "bundle")!

fileprivate let html = """
<!DOCTYPE html>
<html>
    <head>
        <meta name="viewport" content="width=device-width">
        <link rel="stylesheet" href="katex.min.css">
        <script src="katex.min.js"></script>
        <style type="text/css">
         body { background: transparent; }
         #math { float: left; }
        </style>
    </head>
    <body>
        <span id="math"></span>
    </body>
    <script>
     function getMathElement() {
         return document.getElementById('math');
     }
     function render(math, displayMode) {
         try {
             katex.render(math, getMathElement(), {
                 output: 'html',
                 displayMode: displayMode
             });
             return getBounds();
         } catch (error) {
             return error.toString();
         }
     }
     function setColor(color) {
         getMathElement().style.color = color;
     }
     function setNoWrap(noWrap) {
         getMathElement().style.whiteSpace = noWrap ? 'nowrap' : 'unset';
     }
     function getBounds() {
         return getMathElement().getBoundingClientRect().toJSON();
     }
     function loadAllFonts() {
         const fontLoadingPromises = [];
         for (const font of document.fonts) {
             fontLoadingPromises.push(font.load());
         }
         Promise.all(fontLoadingPromises).then(function() {
             window.webkit.messageHandlers.ready.postMessage('ready');
         });
     }
     loadAllFonts();
    </script>
</html>
"""

enum TexError : Error {
    case engineError(message: String)
    case executionError
    case pngConversion
    case concurrentRequest
}

@available(iOS 11.0, *)
class TexRenderer : NSObject, WKScriptMessageHandler {

    lazy var webView = initWebView()
    var ready = false
    var readyListener : (() -> Void)?
    var busy = false

    private func initWebView() -> WKWebView {
        let controller = WKUserContentController()
        controller.add(self, name: "ready")
        controller.add(self, name: "debug")

        let config = WKWebViewConfiguration()
        config.userContentController = controller

        let webView = WKWebView(frame: UIScreen.main.bounds, configuration: config)
        webView.isOpaque = false
        return webView
    }

    private func whenReady(_ completionHandler: @escaping () -> Void) {
        if ready {
            completionHandler()
        } else {
            readyListener = completionHandler
            webView.loadHTMLString(html, baseURL: assetsUrl)
        }
    }

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        switch message.name {
        case "ready":
            ready = true
            readyListener?()
            readyListener = nil
        default:
            debugPrint("Message from WebView to: \(message.name); body: \(message.body)")
        }
    }

    func render(_ math: String, displayMode: Bool, color: String, maxWidth: Double, completionHandler: @escaping (Data?, Error?) -> Void) {
        whenReady { [weak self] in
            guard let self = self else {
                return
            }
            guard !self.busy else {
                completionHandler(nil, TexError.concurrentRequest)
                return
            }
            self.setFrameWidth(maxWidth)
            let escapedMath = math.replacingOccurrences(of: "\\", with: "\\\\")
            let js = "setNoWrap(\(maxWidth.isInfinite)); setColor('\(color)'); render('\(escapedMath)', \(displayMode));"
            log("Executing JavaScript: \(js)")
            self.busy = true
            self.webView.evaluateJavaScript(js) { [weak self] result, error in
                guard let self = self else {
                    return
                }
                if let result = result as? [String:Any] {
                    // Success
                    self.takeSnapshot(result, completionHandler)
                } else if let result = result as? String {
                    // Engine error
                    completionHandler(nil, TexError.engineError(message: result))
                } else {
                    // Other error
                    completionHandler(nil, TexError.executionError)
                }
                self.busy = false
            }
        }
    }

    private func setFrameWidth(_ newWidth: Double) {
        let frameWidth = newWidth.isFinite ? newWidth : Double(UIScreen.main.bounds.width)
        let newFrame = CGRect(x: 0, y: 0, width: Int(frameWidth.rounded(.down)), height: Int(webView.frame.height))
        if webView.frame != newFrame {
            log("New frame width: \(frameWidth); was \(webView.frame.width)")
            webView.frame = newFrame
        }
    }

    private func takeSnapshot(_ bbox: [String:Any], _ completionHandler: @escaping (Data?, Error?) -> Void) {
        let x = (bbox["left"] as! NSNumber).intValue
        let y = (bbox["top"] as! NSNumber).intValue
        let w = (bbox["width"] as! NSNumber).intValue
        let h = (bbox["height"] as! NSNumber).intValue
        let rect = CGRect(x: x, y: y, width: w, height: h)
        log("Taking snapshot of \(rect)")

        let snapConfig = WKSnapshotConfiguration()
        snapConfig.rect = rect
        if #available(iOS 13.0, *) {
            snapConfig.afterScreenUpdates = true
        }
        webView.takeSnapshot(with: snapConfig) { image, error in
            guard let image = image else {
                completionHandler(nil, error)
                return
            }
            guard let data = image.pngData() else {
                completionHandler(nil, TexError.pngConversion)
                return
            }
            completionHandler(data, nil)
        }
    }
}

File: android/src/main/kotlin/com/madlonkay/flutter_tex_js/TexRenderer.kt
package com.madlonkay.flutter_tex_js

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.os.Build
import android.util.Log
import android.webkit.JavascriptInterface
import android.webkit.WebView
import androidx.annotation.MainThread
import kotlinx.coroutines.*
import java.io.ByteArrayOutputStream
import kotlin.math.roundToInt

private const val html =
"""
<!DOCTYPE html>
<html>
    <head>
        <meta name="viewport" content="width=device-width">
        <link rel="stylesheet" href="katex/katex.min.css">
        <script src="katex/katex.min.js"></script>
        <style type="text/css">
         body { background: transparent; margin: 0; }
         #math { float: left; }
        </style>
    </head>
    <body>
        <span id="math"></span>
    </body>
    <script>
     function getMathElement() {
         return document.getElementById('math');
     }
     function render(math, displayMode) {
         try {
             katex.render(math, getMathElement(), {
                 output: 'html',
                 displayMode: displayMode
             });
             // Returning immediately or setting 0 timeout here yielded bad results (incomplete
             // rendering, that weird zooming thing)
             setTimeout(sendBounds, 100);
             return true;
         } catch (error) {
             sendError(error);
             return false;
         }
     }
     function setColor(color) {
         getMathElement().style.color = color;
     }
     function setNoWrap(noWrap) {
         getMathElement().style.whiteSpace = noWrap ? 'nowrap' : 'unset';
     }
     function sendError(error) {
         TexRenderer.onError(error.toString());
     }
     function sendBounds() {
         const bounds = getMathElement().getBoundingClientRect();
         TexRenderer.takeSnapshot(bounds.x, bounds.y, bounds.width, bounds.height);
     }
     function loadAllFonts() {
         const fontLoadingPromises = [];
         for (const font of document.fonts) {
             fontLoadingPromises.push(font.load());
         }
         Promise.all(fontLoadingPromises).then(function() {
             TexRenderer.onReady();
         });
     }
     loadAllFonts();
    </script>
</html>
"""

@SuppressLint("SetJavaScriptEnabled", "AddJavascriptInterface")
class TexRenderer(private val context: Context) : CoroutineScope by MainScope() {

    private val density get() = context.resources.displayMetrics.density
    private val webView: WebView by lazy {
        WebView(context).apply {
            settings.javaScriptEnabled = true
            addJavascriptInterface(this@TexRenderer, "TexRenderer")
            val metrics = context.resources.displayMetrics
            layout(0, 0, metrics.widthPixels, metrics.heightPixels)
            setBackgroundColor(Color.TRANSPARENT)
        }
    }
    private var ready = false
    private var readyListener: (suspend () -> Unit)? = null
    private var resultListener: ((ByteArray?, TexRenderError?) -> Unit)? = null

    @MainThread
    private suspend fun whenReady(completionHandler: suspend () -> Unit) = withContext(Dispatchers.Main) {
        if (ready) {
            completionHandler()
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                WebView.setWebContentsDebuggingEnabled(true)
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                WebView.enableSlowWholeDocumentDraw()
            }
            readyListener = completionHandler
            webView.loadDataWithBaseURL("file:///android_asset/", html, "text/html", null, null)
        }
    }

    @JavascriptInterface
    fun onReady() {
        Log.d("AMK", "onReady called; thread=${Thread.currentThread()}")
        ready = true
        val listener = readyListener!!
        launch {
            listener.invoke()
        }
        readyListener = null
    }

    @JavascriptInterface
    fun onError(error: String) {
        Log.d("AMK", "onError called; thread=${Thread.currentThread()}")
        val listener = resultListener!!
        launch {
            val err = TexRenderError("RenderError", "An error occurred during rendering", error)
            listener(null, err)
        }
        resultListener = null
    }

    @JavascriptInterface
    fun takeSnapshot(x: Double, y: Double, width: Double, height: Double) {
        val xPx = (x * density).roundToInt()
        val yPx = (y * density).roundToInt()
        val widthPx = (width * density).roundToInt()
        val heightPx = (height * density).roundToInt()
        Log.d("AMK", "Taking snapshot of [$x, $y, $width, $height], scaled to [$xPx, $yPx, $widthPx, $heightPx]")
        var bitmap = Bitmap.createBitmap(xPx + widthPx, yPx + heightPx, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        val listener = resultListener!!
        launch {
            webView.draw(canvas)
            if (xPx != 0 || yPx != 0) {
                bitmap = Bitmap.createBitmap(bitmap, xPx, yPx, widthPx, heightPx)
            }
            val bytes = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, bytes)
            listener.invoke(bytes.toByteArray(), null)
        }
        resultListener = null
    }

    @MainThread
    suspend fun render(math: String, displayMode: Boolean, color: String, maxWidth: Double, completionHandler: (ByteArray?, TexRenderError?) -> Unit) = withContext(Dispatchers.Main) {
        whenReady {
            if (resultListener != null) {
                val err = TexRenderError("ConcurrencyError", "A render job was already in progress", null)
                completionHandler(null, err)
                return@whenReady
            }
            resultListener = completionHandler
            val js = "setNoWrap(${maxWidth.isInfinite()}); setColor('$color'); render('$math', $displayMode);"
            Log.d("AMK", "Executing JavaScript: $js")
            setViewWidth(maxWidth)
            // We call loadUrl instead of evaluateJavascript for backwards compatibility, and
            // because we can't really make use of the immediately returned result
            webView.loadUrl("javascript:$js")
        }
    }

    @MainThread
    private fun setViewWidth(width: Double) {
        val newWidth = if (width.isFinite()) {
            (width * density).roundToInt()
        } else {
            context.resources.displayMetrics.widthPixels
        }
        if (webView.width != newWidth) {
            Log.d("AMK", "New frame width: $newWidth; was ${webView.width}")
            webView.layout(0, 0, newWidth, webView.height)
        }
    }
}

data class TexRenderError(val code: String, val message: String?, val details: String?)
